From 309102c868d0c687acf91ddea7581e12f24bb7fc Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Fri, 25 Jul 2025 11:54:18 +0200
Subject: [PATCH 904/954] feat(stm32mp2): use SCMI power state protocol for
 CM33 boot

Use the SCMI power state protocol to advertise TF-M of
shutdown reset and system suspend request.

The SCMI driver is initialized with PSCI service.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git aea85369d5d268cc724d85914aea68ec7489a940 feat(stm32mp2): use SCMI power state protocol for CM33 boot]
Change-Id: Ia8d0e1b830eaa3b262f92f3a2def032c29fce044
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/475540
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Tested-by: Maxime MERE <maxime.mere@foss.st.com>
Domain-Review: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/reset/stm32mp2_reset.c             | 12 ++-
 include/drivers/st/stm32mp_reset.h            |  1 +
 plat/st/stm32mp2/platform.mk                  |  3 +-
 .../stm32mp2/{services => scmi}/scmi_common.c | 98 +++++++++++++++++--
 .../{services => scmi}/scmi_private.h         |  0
 plat/st/stm32mp2/services/ca35ss_clk_svc.c    |  5 -
 plat/st/stm32mp2/stm32mp2_pm.c                | 30 +++++-
 7 files changed, 127 insertions(+), 22 deletions(-)
 rename plat/st/stm32mp2/{services => scmi}/scmi_common.c (83%)
 rename plat/st/stm32mp2/{services => scmi}/scmi_private.h (100%)

diff --git a/drivers/st/reset/stm32mp2_reset.c b/drivers/st/reset/stm32mp2_reset.c
index bf557d4fd3..4f1e73b038 100644
--- a/drivers/st/reset/stm32mp2_reset.c
+++ b/drivers/st/reset/stm32mp2_reset.c
@@ -63,19 +63,21 @@ int stm32mp_reset_deassert(uint32_t id, unsigned int to_us)
 	return reset_toggle(id, to_us, false);
 }
 
+#if !STM32MP_M33_TDCID
 void __dead2 stm32mp_system_reset(void)
 {
-#if STM32MP_M33_TDCID
-	INFO("BL31: System reset, waiting watchdog\n");
-	/* TODO: use SCMI channel to request C1 reset to cortex M33 */
-#else
 	uintptr_t rcc_base = stm32mp_rcc_base();
 
 	mmio_setbits_32(rcc_base + RCC_GRSTCSETR, RCC_GRSTCSETR_SYSRST);
-#endif
 
 	/* Loop in case system reset is not immediately caught */
 	for ( ; ; ) {
 		;
 	}
 }
+
+void __dead2 stm32mp_system_cold_reset(void)
+{
+	stm32mp_system_reset();
+}
+#endif
diff --git a/include/drivers/st/stm32mp_reset.h b/include/drivers/st/stm32mp_reset.h
index 40ecdd59ab..bd4f01f7cb 100644
--- a/include/drivers/st/stm32mp_reset.h
+++ b/include/drivers/st/stm32mp_reset.h
@@ -51,5 +51,6 @@ static inline void stm32mp_reset_release(uint32_t reset_id)
  * Manage system reset control
  */
 void __dead2 stm32mp_system_reset(void);
+void __dead2 stm32mp_system_cold_reset(void);
 
 #endif /* STM32MP_RESET_H */
diff --git a/plat/st/stm32mp2/platform.mk b/plat/st/stm32mp2/platform.mk
index a0227e791c..ce1974f013 100644
--- a/plat/st/stm32mp2/platform.mk
+++ b/plat/st/stm32mp2/platform.mk
@@ -280,6 +280,7 @@ TF_CFLAGS			+=	-mbranch-protection=none
 
 # Include paths and source files
 PLAT_INCLUDES			+=	-Iplat/st/stm32mp2/include/
+PLAT_INCLUDES			+=	-Iplat/st/stm32mp2/scmi/
 PLAT_INCLUDES			+=	-Idrivers/st/ddr/phy/phyinit/include/
 PLAT_INCLUDES			+=	-Idrivers/st/ddr/phy/firmware/include/
 
@@ -438,7 +439,7 @@ BL31_SOURCES			+=	plat/st/common/stm32mp_svc_setup.c			\
 BL31_SOURCES			+=	services/arm_arch_svc/arm_arch_svc_setup.c
 
 ifeq (${STM32MP_M33_TDCID},1)
-BL31_SOURCES			+=	plat/st/stm32mp2/services/scmi_common.c
+BL31_SOURCES			+=	plat/st/stm32mp2/scmi/scmi_common.c
 endif
 
 ifeq (${STM32MP_SIP_CA33SS_CLK},1)
diff --git a/plat/st/stm32mp2/services/scmi_common.c b/plat/st/stm32mp2/scmi/scmi_common.c
similarity index 83%
rename from plat/st/stm32mp2/services/scmi_common.c
rename to plat/st/stm32mp2/scmi/scmi_common.c
index 880c65bc7a..c9f340e939 100644
--- a/plat/st/stm32mp2/services/scmi_common.c
+++ b/plat/st/stm32mp2/scmi/scmi_common.c
@@ -39,6 +39,13 @@
 #define SCMI_MSG_PROTOCOL_VERSION	0x0U
 #define SCMI_MSG_PROTOCOL_MSG_ATTR	0x2U
 
+/* SCMI Discovery base messages (based on drivers/scmi-msg/base.h) */
+#define SCMI_BASE_DISCOVER_VENDOR			0x03U
+#define SCMI_BASE_DISCOVER_SUB_VENDOR			0x04U
+#define SCMI_BASE_DISCOVER_IMPLEMENTATION_VERSION	0x05U
+#define SCMI_BASE_DISCOVER_LIST_PROTOCOLS		0x06U
+#define SCMI_BASE_DISCOVER_AGENT			0x07U
+
 /* SCMI voltage domain protocol message IDs */
 #define SCMI_MSG_VOLTAGE_LEVEL_SET	0x7U
 #define SCMI_MSG_VOLTAGE_LEVEL_GET	0x8U
@@ -74,6 +81,8 @@
 #define SCMI_MAX_MESSAGE_PAYLOAD_SIZE	((SMT_BUF_SLOT_SIZE - SCMI_SHMEM_PAYLOAD_OFFSET) \
 					/ sizeof(uint32_t))
 
+#define SIZE_NO_CHECK			((uint32_t)-1)
+
 struct scmi_shmem_layout {
 	uint32_t reserved0;
 	uint32_t channel_status;
@@ -192,7 +201,8 @@ static int32_t scmi_rsp_check(uint32_t size)
 	if (shmem->length == 2 * sizeof(uint32_t)) {
 		/* Return the first paylod value = status. For example, NOT SUPPORTED */
 		return (int32_t)shmem->payload[0];
-	} else if (shmem->length != ((1 + size) * sizeof(uint32_t))) {
+	} else if (size != SIZE_NO_CHECK &&
+		   shmem->length != ((1 + size) * sizeof(uint32_t))) {
 		INFO("SCMI 0x%x msg %u: size %u, expected %lu\n",
 		     SCMI_MSG_GET_PROTOCOL(shmem->message_header),
 		     SCMI_MSG_GET_MSG_ID(shmem->message_header),
@@ -259,9 +269,6 @@ static int32_t smci_proto_version_check(uint32_t proto_id, uint32_t expected)
 			     proto_id, version, expected);
 			return SCMI_OUT_OF_RANGE;
 		}
-	} else if (ret == SCMI_NOT_SUPPORTED) {
-		/* Not supported is a valid error */
-		ret = SCMI_SUCCESS;
 	} else {
 		ERROR("SCMI protocol 0x%x: init error %d\n", proto_id, ret);
 	}
@@ -415,6 +422,36 @@ int32_t scmi_sys_pwr_state_set(uint32_t flags, uint32_t system_state)
 	return (int32_t)shmem->payload[0]; /* status */
 }
 
+/*
+ * API to set the SCMI protocl list with BASE_DISCOVER_LIST_PROTOCOLS
+ */
+int32_t scmi_base_discover_list_proto(uint32_t *num, uint8_t *protocols, uint8_t size)
+{
+	int32_t ret;
+
+	ret = scmi_channel_prepare(SCMI_PROTOCOL_ID_BASE, SCMI_BASE_DISCOVER_LIST_PROTOCOLS, 1U);
+	if (ret != SCMI_SUCCESS) {
+		return ret;
+	}
+
+	shmem->payload[0] = 0U; /* skip */
+
+	scmi_ring_doorbell();
+
+	ret = scmi_rsp_wait(TIMEOUT_10MS_IN_US, SIZE_NO_CHECK);
+	if (ret != SCMI_SUCCESS) {
+		return ret;
+	}
+
+	*num = shmem->payload[1];
+	if (*num > size) {
+		*num = size;
+	}
+	memcpy(protocols, &shmem->payload[2], *num);
+
+	return (int32_t)shmem->payload[0]; /* status */
+}
+
 /* Check if the message SCMI SYSTEM_POWER_STATE_SET is supported */
 static void scmi_check_sys_pwr_state_set(void)
 {
@@ -424,30 +461,71 @@ static void scmi_check_sys_pwr_state_set(void)
 	ret = scmi_proto_msg_attr(SCMI_PROTOCOL_ID_SYS_POWER, SCMI_MSG_SYS_PWR_STATE_SET, &attr);
 	if (ret == SCMI_SUCCESS) {
 		scmi_sys_pwr_state_set_supported = true;
-	} else {
-		scmi_sys_pwr_state_set_supported = false;
-		INFO("SCMI SYSTEM_POWER_STATE_SET not supported (%d)\n", ret);
 	}
 }
 
 int32_t scmi_init(void)
 {
 	int32_t ret;
+	uint8_t proto[8];
+	uint32_t nb_proto;
+	uint32_t i;
 
 	scmi_sys_pwr_state_set_supported = false;
 
 	/* Initialize channel */
 	scmi_channel_clear();
 
-	/* Check supported protocol and messages */
-	ret = smci_proto_version_check(SCMI_PROTOCOL_ID_SYS_POWER, SCMI_PROTO_VER_SYS_PWR);
+	/* Get Liste of supported ptotocols */
+	ret = scmi_base_discover_list_proto(&nb_proto, proto, sizeof(proto));
 	if (ret != SCMI_SUCCESS) {
+		INFO("SCMI init failed(%d)\n", ret);
 		goto err;
 	}
-	scmi_check_sys_pwr_state_set();
+
+	/* Check supported protocol and messages */
+	for (i = 0U; i < nb_proto; i++) {
+		if (proto[i] != SCMI_PROTOCOL_ID_SYS_POWER) {
+			continue;
+		}
+		ret = smci_proto_version_check(SCMI_PROTOCOL_ID_SYS_POWER, SCMI_PROTO_VER_SYS_PWR);
+		if (ret != SCMI_SUCCESS) {
+			goto err;
+		}
+		scmi_check_sys_pwr_state_set();
+		break;
+	}
 
 	ret = smci_proto_version_check(SCMI_PROTOCOL_ID_VOLTAGE_DOMAIN, SCMI_PROTO_VER_VOLTAGE);
 
 err:
 	return ret;
 }
+
+void __dead2 stm32mp_system_reset(void)
+{
+	int ret = scmi_sys_pwr_state_set(SCMI_SYS_PWR_GRACEFUL_REQ, SCMI_SYS_PWR_WARM_RESET);
+
+	if (ret != SCMI_SUCCESS) {
+		INFO("SCMI SYSTEM_POWER_STATE_SET error (%d), waiting watchdog\n", ret);
+	}
+
+	/* Loop in case system reset is not immediately caught */
+	for ( ; ; ) {
+		;
+	}
+}
+
+void __dead2 stm32mp_system_cold_reset(void)
+{
+	int ret = scmi_sys_pwr_state_set(SCMI_SYS_PWR_GRACEFUL_REQ, SCMI_SYS_PWR_COLD_RESET);
+
+	if (ret != SCMI_SUCCESS) {
+		INFO("SCMI SYSTEM_POWER_STATE_SET error (%d), waiting watchdog\n", ret);
+	}
+
+	/* Loop in case system reset is not immediately caught */
+	for ( ; ; ) {
+		;
+	}
+}
diff --git a/plat/st/stm32mp2/services/scmi_private.h b/plat/st/stm32mp2/scmi/scmi_private.h
similarity index 100%
rename from plat/st/stm32mp2/services/scmi_private.h
rename to plat/st/stm32mp2/scmi/scmi_private.h
diff --git a/plat/st/stm32mp2/services/ca35ss_clk_svc.c b/plat/st/stm32mp2/services/ca35ss_clk_svc.c
index e07fc235c0..b5a7121070 100644
--- a/plat/st/stm32mp2/services/ca35ss_clk_svc.c
+++ b/plat/st/stm32mp2/services/ca35ss_clk_svc.c
@@ -438,11 +438,6 @@ uint32_t ca35ss_clk_svc_setup(void)
 		}
 	}
 
-	if (scmi_init() != 0) {
-		ERROR("%s: couldn't get SCMI voltd protocol\n", __func__);
-		return STM32_SMC_FAILED;
-	}
-
 	/* Verify SCMI voltage domain availability */
 	if (scmi_voltd_protocol_version(&voltd_version) != SCMI_SUCCESS) {
 		ERROR("%s: couldn't get SCMI voltd protocol\n", __func__);
diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index 238870d018..3ac1a0982a 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -32,6 +32,7 @@
 #include <stm32mp2_private.h>
 
 #include "../../../lib/psci/psci_private.h"
+#include "scmi_private.h"
 
 /* Default value for STM32MP25 with STPMIC25, defined in AN5727 */
 #define DEFAULT_POPL_D1		3U
@@ -474,6 +475,13 @@ static void print_mode_verbose(const char *mode)
 	VERBOSE("Entering %s low power mode\n", mode);
 }
 
+static void stm32_set_scmi_sys_pwr(uint32_t system_state)
+{
+#if STM32MP_M33_TDCID
+	scmi_sys_pwr_state_set(SCMI_SYS_PWR_GRACEFUL_REQ, system_state);
+#endif
+}
+
 static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 {
 	uintptr_t pwr_base = stm32mp_pwr_base();
@@ -593,6 +601,7 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 		stm32mp_gic_cpuif_disable();
 		stm32mp_gic_save();
 		stm32mp2_pll1_disable();
+		stm32_set_scmi_sys_pwr(SCMI_SYS_PWR_SUSPEND);
 		break;
 
 #if !STM32MP_M33_TDCID
@@ -635,6 +644,7 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 		stm32mp_gic_save();
 #endif
 		stm32mp2_pll1_disable();
+		stm32_set_scmi_sys_pwr(SCMI_SYS_PWR_SUSPEND);
 		break;
 
 	default:
@@ -1020,7 +1030,7 @@ static void __dead2 stm32_system_off(void)
 	/* Disable STATE_RUNNING state for this core */
 	stm32mp_state_set(core_id, STATE_RUNNING, false);
 
-	VERBOSE("BL31: power off\n");
+	INFO("BL31: power off\n");
 
 	dsb();
 	isb();
@@ -1032,7 +1042,22 @@ static void __dead2 stm32_system_off(void)
 
 static void __dead2 stm32_system_reset(void)
 {
+	INFO("BL31: System cold reset\n");
+	stm32mp_system_cold_reset();
+}
+
+static int stm32_system_reset2(int is_vendor, int reset_type, u_register_t cookie)
+{
+	if (is_vendor || (reset_type != PSCI_RESET2_SYSTEM_WARM_RESET))
+		return PSCI_E_INVALID_PARAMS;
+
+	INFO("BL31: System reset\n");
 	stm32mp_system_reset();
+
+	/* This shouldn't be reached */
+	panic();
+
+	return 0;
 }
 
 /**
@@ -1250,6 +1275,7 @@ static const plat_psci_ops_t stm32_psci_ops = {
 	.pwr_domain_pwr_down_wfi = stm32_pwr_domain_pwr_down_wfi,
 	.system_off = stm32_system_off,
 	.system_reset = stm32_system_reset,
+	.system_reset2 = stm32_system_reset2,
 	.validate_power_state = stm32_validate_power_state,
 	.validate_ns_entrypoint = stm32_validate_ns_entrypoint,
 	.get_sys_suspend_power_state = stm32_get_sys_suspend_power_state,
@@ -1433,6 +1459,8 @@ static void stm32_pm_init(void *fdt)
 
 #if !STM32MP_M33_TDCID
 	stm32_pm_tdcid_init(fdt);
+#else
+	scmi_init();
 #endif
 }
 
-- 
2.53.0

